satpy/readers/netcdf_utils.py

```diff
@@ -349,3 +377,56 @@ def _compose_replacement_names(variable_name_replacements, var, variable_names):
         for val in vals:
             if key in var:
                 variable_names.append(var.format(**{key: val}))
+
+
+class NetCDF4FsspecFileHandler(NetCDF4FileHandler):
+    """NetCDF4 file handler using fsspec to read files remotely."""
+
+    def _get_file_handle(self):
+        try:
+            # Default to using NetCDF4 backend for local files
+            return super()._get_file_handle()
+        except OSError:
+            # The netCDF4 lib raises either FileNotFoundError or OSError for remote files. OSError catches both.
+            import h5netcdf
+            f_obj = open_file_or_filename(self.filename)
+            self._use_h5netcdf = True
+            return h5netcdf.File(f_obj, 'r')
```

Comment:
This seems prone to failure or unexpected results. I've recently discovered that depending on the build of NetCDF-C, netcdf can handle different types of files and URLs. It also sometimes depends on if `#mode=bytes` is appended to the URL. So what I'm saying is maybe there are sometimes when the NetCDF C library will be able to handle remote files and this file handler won't be used. Are we OK with that? Should we be more explicit with `if self.filename.startswith("s3://")` or something?

Is there anything special being done in this new file handler that should also be in the parent file handler?

@mraspaud said:

> The way the fsspec-based file handler is done feels a bit hacked, but unfortunately the way the reader works doesn't allow polymorphism here, so nothing to do here.

I'm not sure I understand. Is there a way to put this try/except and basically all off this logic in the main parent class?

Reply:
> @mraspaud said:
> 
> > The way the fsspec-based file handler is done feels a bit hacked, but unfortunately the way the reader works doesn't allow polymorphism here, so nothing to do here.
> 
> I'm not sure I understand. Is there a way to put this try/except and basically all off this logic in the main parent class?

No, that's the problem. I would have liked satpy internals to be able to choose between two classes depending eg on the uri, but the way it works now we can't. If we had an interface like xarray's open_dataset (or open_datatree) for that matter, that function would be able to instantiate the right class, but now we hard code the filehandler class in the yaml file.

Reply:
This may be more work than this PR can support right now, but can we:

Split the base class and this new class into 3 total classes. Essentially a "netcdf4 handler" and an "h5netcdf file handler" and then a single class that uses both of these classes as "mixins"? Or maybe just sub-objects? The creation of the object can be determined when the file handler tries to open `self.filename`. It still has the try/except but instead of calling NetCDF4-python directly it creates an instance of the "netcdf4 handler" class and if exception then it makes the h5netcdf handler class. They would be essentially two classes implementing the file-reading interfaces that the Satpy NetCDF4 utility handler needs to get attributes/data.

Reply:
Indeed, that would work, but probably something for another PR.

Reply:
I'm OK removing this new class, if the NetCDF4 library some day _always_ supports the remote reading. Relying on a specific source (`conda-forge`, for example) for the library won't be enough, as not everybody can use it.

I'm open to ideas how the `.startswith("s3://")` could be enhanced.

Reply:
The `try/except` might be the best option. I think we're generally OK with the behavior of "do whatever you have to do to read this file", but we'll have to keep it in mind when debugging things for us and user bug reports ("wait, do you only have h5netcdf installed?").

The only reason *not* to do the try/except is that NetCDF4-C may be able to read URLs (HTTP for sure, s3:// some times I think) but may not perform as well as h5netcdf depending on what modes/methods are being used to read the data (as a stream of bytes versus HTTP byte range requests). The worst part about this would be that it would be hard to detect because things would still be working, just slower.

I think my overall vote is for merging this stuff into one user-level class (likely constructed from 2 or more other classes) and leaving the try/except.

Reply:
I created a new issue for this: https://github.com/pytroll/satpy/issues/2362